### PR TITLE
equinix: set default deny iptables policy

### PR DESCRIPTION
**Pull request by displague**
This PR applies changes for the Equinix provider:
* sets the default firewall policy to drop (follow-up to https://github.com/juju/juju/pull/13695#issuecomment-1206010419) (borrows from rackspace provider)
* filter flavors that are not available for on_demand server deployments (these plans are returned by default today because they are evaluated, incorrectly, as the best match)
* calculates CPU Core count based on text description in the plan details
* Test coverage is improved for flavor / plan selection

## QA steps

*Commands to run to verify that the change works.*

```sh
git clone https://github.com/juju/juju/
cd juju
git fetch origin pulls/14568/head:equinix-metal-iptables-policy
git checkout equinixmetal-firewaller
go install ./cmd/juju
which juju # should be the one installed in the GOBIN path

YOUR_IP=$(curl -4 icanhazip.com)
YOUR_IPV6=$(curl -6 icanhazip.com)

# One of the fixes in this PR makes it no longer necessary to specify the EM OS image and plan.

juju add-credential equinix
juju bootstrap equinix/da equinix-controller \
  --bootstrap-series=focal \
  --build-agent # to use the local linux environment development build remotely

juju deploy mysql --channel=edge --series=bionic && \
juju deploy wordpress --channel=edge --series=bionic --to=0 && \
juju add-relation wordpress mysql && \
juju expose wordpress --to-cidrs=${YOUR_IP}/32 # your ipv4, add ",${YOUR_IPV6}/128" if also available
# ^ it should now be accessible
# juju status to get the IP and load it in your browser

juju unexpose wordpress
# ^ it should no longer be accessible. verify in your browser.

# debug: if the page loads when it is not exposed, report the following log:
# TODO: how to log helpful debugging
# iptables -L -t filter 
```

### CODE REVIEW

File: provider/equinix/environ.go [hunk 1]
```diff
@@ -277,7 +284,7 @@ func newConfig(cfg, old *config.Config) (*environConfig, error) {
 	return ecfg, nil
 }
 
-func (e *environ) StartInstance(ctx context.ProviderCallContext, args environs.StartInstanceParams) (result *environs.StartInstanceResult, resultErr error) {
+func (e *environ) configureInstance(args environs.StartInstanceParams) (*instances.InstanceSpec, error) {
 	instanceTypes, err := e.InstanceTypes(ctx, constraints.Value{})
```
Comment (manadart) [inline provider/equinix/environ.go:288]: `ctx` is no longer passed here, preventing compilation.

File: provider/equinix/environ.go [hunk 2]
```diff
@@ -17,6 +19,7 @@ import (
 	"github.com/juju/loggo"
 	"github.com/juju/retry"
 	"github.com/juju/schema"
+	"github.com/juju/utils/ssh"
```
Comment (manadart) [inline provider/equinix/environ.go:22]: Should be `github.com/juju/utils/v3/ssh`

Comment (displague) [reply provider/equinix/environ.go]: https://github.com/juju/juju/compare/9ce32f0f7b0236793d0771b8cdae0dfbd6207b6f..3a6cdc2210c23104b1b3cf867dce9907035060e3

**Review by manadart — COMMENTED** [review]
There's a panic when bootstrapping with this patch:
```
panic: regexp: Compile(`(\d+)[ -]core)`): error parsing regexp: unexpected ): `(\d+)[ -]core)`

goroutine 1 [running]:
regexp.MustCompile({0x45e1876, 0xe})
        /usr/local/go/src/regexp/regexp.go:319 +0xbb
github.com/juju/juju/provider/equinix.(*environ).supportedInstanceTypes(0xc000082c00)
        /home/joseph/go/src/github.com/juju/juju/provider/equinix/environ.go:719 +0x52a
github.com/juju/juju/provider/equinix.(*environ).InstanceTypes(0xc0005b5c90?, {0x0?, 0xc0005b5cd0?}, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/equinix/environ.go:168 +0x5e
github.com/juju/juju/provider/equinix.(*environ).configureInstance(_, {_, _}, {{0xc000298600, 0x24}, {0xc000faa380, 0x0, 0x0, 0x0, 0x0, ...}, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/equinix/environ.go:288 +0x65
github.com/juju/juju/provider/equinix.(*environ).StartInstance(_, {_, _}, {{0xc000298600, 0x24}, {0xc000faa380, 0x0, 0x0, 0x0, 0x0, ...}, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/equinix/environ.go:421 +0x7a
github.com/juju/juju/provider/common.BootstrapInstance({_, _}, {_, _}, {_, _}, {{0xc000ce8618, 0x7}, {0xc000ce8dae, 0x2}, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/common/bootstrap.go:265 +0x122f
github.com/juju/juju/provider/common.Bootstrap({_, _}, {_, _}, {_, _}, {{0xc000ce8618, 0x7}, {0xc000ce8dae, 0x2}, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/common/bootstrap.go:58 +0x9e
github.com/juju/juju/provider/equinix.(*environ).Bootstrap(_, {_, _}, {_, _}, {{0xc000ce8618, 0x7}, {0xc000ce8dae, 0x2}, 0xc0002f3560, ...})
        /home/joseph/go/src/github.com/juju/juju/provider/equinix/environ.go:81 +0x93
github.com/juju/juju/environs/bootstrap.bootstrapIAAS({_, _}, {_, _}, {_, _}, {{0x0, 0x0, 0x0, 0x0, ...}, ...}, ...)
        /home/joseph/go/src/github.com/juju/juju/environs/bootstrap/bootstrap.go:583 +0x1b3e
github.com/juju/juju/environs/bootstrap.Bootstrap({_, _}, {_, _}, {_, _}, {{0x0, 0x0, 0x0, 0x0, ...}, ...})
        /home/joseph/go/src/github.com/juju/juju/environs/bootstrap/bootstrap.go:708 +0x2a9
github.com/juju/juju/cmd/juju/commands.bootstrapFuncs.Bootstrap(...)
        /home/joseph/go/src/github.com/juju/juju/cmd/juju/commands/bootstrap.go:451
github.com/juju/juju/cmd/juju/commands.(*bootstrapCommand).Run(0xc00093c380, 0xc00059ec60)
        /home/joseph/go/src/github.com/juju/juju/cmd/juju/commands/bootstrap.go:978 +0x2909
github.com/juju/juju/cmd/modelcmd.(*modelCommandWrapper).Run(0xc0005301e0, 0xc00088cba0?)
        /home/joseph/go/src/github.com/juju/juju/cmd/modelcmd/modelcommand.go:663 +0x123
github.com/juju/juju/cmd/modelcmd.(*baseCommandWrapper).Run(0xc0003f0540, 0x200000006?)
        /home/joseph/go/src/github.com/juju/juju/cmd/modelcmd/base.go:551 +0xaf
github.com/juju/cmd/v3.(*SuperCommand).Run(0xc000912500, 0xc00059ec60)
        /home/joseph/go/src/github.com/juju/juju/vendor/github.com/juju/cmd/v3/supercommand.go:523 +0x371
github.com/juju/cmd/v3.Main({0x4dbfa78, 0xc000912500}, 0xc00059ec60, {0xc0007921b0, 0x9, 0x9})
        /home/joseph/go/src/github.com/juju/juju/vendor/github.com/juju/cmd/v3/cmd.go:397 +0x24b
github.com/juju/juju/cmd/juju/commands.jujuMain.Run({0xc0000ec140?}, {0xc000000140, 0xa, 0xa})
        /home/joseph/go/src/github.com/juju/juju/cmd/juju/commands/main.go:199 +0x80f
github.com/juju/juju/cmd/juju/commands.Main(...)
        /home/joseph/go/src/github.com/juju/juju/cmd/juju/commands/main.go:122
main.main()
        /home/joseph/go/src/github.com/juju/juju/cmd/juju/main.go:27 +0x72
```

**Comment by manadart** [comment]
There's a problem here.

Upon bootstrap, I had to answer the host verification prompt for the SSH attempt. The real issue though, is that new machines can't be provisioned due to the same.

Status of added machine:
```
failed to start machine 0 (waited for 5m0s without being able to connect: Host key verification failed.), retrying in 10s (10 more attempts)
```

**Comment by manadart** [comment]
In fact, you already inject cloud-init scripts to manipulate ip-tables rules (in the script run by cron). You could add the default deny/drop behaviour as a new script.

**Comment by displague** [comment]
Some notes from @manadart:
* A problem I experienced testing this branch is that machines were being reprovisioned out of control. manadart identified that this is due to the use of the SSH provisioning steps while the machine is not ready. Take advantage of the cloud-init iptables modifications to set DROP policy on INPUT.
* The cron job that resets the iptables rules for fan can be removed now that iptables-persist is installed.

File: provider/equinix/environ.go [hunk 3]
```diff
@@ -302,20 +307,38 @@ func (e *environ) StartInstance(ctx context.ProviderCallContext, args environs.S
...
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	cloudCfg.AddPackage("iptables-persistent")
+
+	// Set a default INPUT policy of drop, permitting ssh
+	iptablesDefault := []string{
+		"iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT",
+		"iptables -A INPUT -m state --state ESTABLISHED,RELATED -j ACCEPT",
+		"iptables -A INPUT -p tcp --dport ssh -j ACCEPT",
```
Comment (displague) [inline provider/equinix/environ.go:324]: My rules may be insufficient:
```
Setting up dns-root-data (2019052802) ...
Setting up libidn11:amd64 (1.33-2.2ubuntu2) ...
Setting up bridge-utils (1.6-2ubuntu1) ...
Setting up dnsmasq-base (2.80-1.1ubuntu1.5) ...
Setting up ubuntu-fan (0.12.13ubuntu0.1) ...
Created symlink /etc/systemd/system/multi-user.target.wants/ubuntu-fan.service → /lib/systemd/system/ubuntu-fan.service.
Processing triggers for systemd (245.4-4ubuntu3.18) ...
Processing triggers for man-db (2.9.1-1) ...
Processing triggers for dbus (1.12.16-2ubuntu2.2) ...
Processing triggers for libc-bin (2.31-0ubuntu9.9) ...
728a5565b2f05e251b14f9f44a695e05f1d38ecc79830ce28d81ce8c99a76939  /var/lib/juju/tools/2.9.35.1-ubuntu-amd64/tools.tar.gz
925d99bcc83bb9accc88df729d12e5c83a07ece73a94f936fbbae959184ac12d  /var/lib/juju/gui/gui.tar.bz2
Installing Juju machine agent
2022-10-11 17:06:10 INFO juju.cmd supercommand.go:56 running jujud [2.9.35.1  gc go1.18.4]
2022-10-11 17:06:11 INFO juju.agent identity.go:22 writing system identity file
2022-10-11 17:06:11 WARNING juju.mongo mongo.go:481 overwriting args.dataDir (set to /var/lib/juju) to /var/snap/juju-db/common
2022-10-11 17:06:11 INFO juju.mongo mongo.go:486 Ensuring mongo server is running; data directory /var/snap/juju-db/common; port 37017
2022-10-11 17:06:11 WARNING juju.mongo service.go:323 configuring mongod  with --noauth flag enabled
2022-10-11 17:06:11 INFO juju.packaging manager.go:103 installing "juju-db" via "snap"
2022-10-11 17:06:11 INFO juju.packaging.manager run.go:88 Running: snap install  --channel 4.4/stable juju-db
2022-10-11 17:12:55 INFO juju.worker.peergrouper initiate.go:68 finished InitiateMongoServer
ERROR failed to start mongo: cannot initiate replica set: cannot dial mongo to initiate replicaset: no reachable servers
ERROR failed to bootstrap model: subprocess encountered error code 1
```

Comment (manadart) [reply provider/equinix/environ.go:324]: I believe for controllers you will need to expose the Mongo port, which can be accessed via:
```go
args.InstanceConfig.ControllerConfig.StatePort()
```

**Comment by displague** [comment]
Thanks, @manadart. I was able to test verify that with the current build, following the PR testing notes, I was able to deploy wordpress, expose it, access it, unexpose it, and no longer access it. The iptables rules were consistent with each change.

**Comment by displague** [comment]
Passing linters now.

**Review by manadart — APPROVED** [review]
All looks good. Approved conditionally, based on removal of dead code.

File: provider/equinix/environ.go [hunk 4]
```diff
@@ -277,7 +279,7 @@ func newConfig(cfg, old *config.Config) (*environConfig, error) {
 	return ecfg, nil
 }
 
-func (e *environ) StartInstance(ctx context.ProviderCallContext, args environs.StartInstanceParams) (result *environs.StartInstanceResult, resultErr error) {
+func (e *environ) configureInstance(ctx context.ProviderCallContext, args environs.StartInstanceParams) (*instances.InstanceSpec, error) {
```
[the remark this answers is not in the document]
Comment (displague) [reply provider/equinix/environ.go:246]: resolved in 8fa1bb8

File: provider/equinix/environ.go (hunk 3 above)
Comment (displague) [reply provider/equinix/environ.go]: The default ruleset is now exposing the following for Controllers:			
```
args.InstanceConfig.ControllerConfig.APIPort(),
args.InstanceConfig.ControllerConfig.StatePort(),
args.InstanceConfig.ControllerConfig.ControllerAPIPort(),
```

File: provider/equinix/environ.go [hunk 5]
```diff
@@ -403,9 +397,25 @@ while true; do
     exit 0
 done
 EOF`,
-		"(crontab -l 2>/dev/null; echo '@reboot bash /root/juju-fixups.sh') | crontab -",
```
Comment (displague) [inline provider/equinix/environ.go:406]: This crontab entry should not have been removed. The persistence that was added for iptables does not persist IP routes.  Perhaps we should run just the `ip routes` section of `juju-fixups.sh` in cron (or persist this in ip-up.d files).

**Pull request merged**